src/Utilities/cachingoptimizer.jl

```diff
@@ -34,19 +34,24 @@ mutable struct CachingOptimizer{MT<:MOI.ModelLike} <: MOI.AbstractOptimizer
     # optimizer indices.
 end
 
-CachingOptimizer(model_cache::MOI.ModelLike, mode::CachingOptimizerMode) = CachingOptimizer(model_cache, nothing, NoOptimizer, mode, IndexMap(), IndexMap())
+function CachingOptimizer(model_cache::MOI.ModelLike, mode::CachingOptimizerMode)
+    CachingOptimizer{Union{Nothing,MOI.AbstractOptimizer}, typeof(model_cache)}(nothing, model_cache, NoOptimizer, mode, IndexMap(), IndexMap())
```

Comment:
`Union{Nothing,MOI.AbstractOptimizer}` is a really unusual type parameter. Why not type the `optimizer` as `Union{Nothing, OptimizerType}`? My understanding is that there's very little overhead in 0.7 for `Union{Nothing, T}` if `T` is concrete.

Reply:
Indeed, I have updated the description which a benchmark showing this